File: Source/OldCode/Archive/2005 TVA Code Library/Version/TVACodeLibrary/TVA.Core_CS/Diagnostics/PerformanceCounter.cs
using System.Diagnostics;
using System.Linq;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Threading;

// 06/04/2007


namespace TVA
{
	namespace Diagnostics
	{
		
		public class PerformanceCounter : IDisposable
		{
			
			
			#region " Member Declaration "
			
			private float m_lastValue;
			private float m_minimumValue;
			private float m_maximumValue;
			private int m_averagingWindow;
			
			private System.Diagnostics.PerformanceCounter m_counter;
			private List<double> m_counterValues;
			
			#endregion
			
			#region " Code Scope: Public "
			
			public const int DefaultAveragingWindow = 120;
			
			public PerformanceCounter(string categoryName, string counterName, string instanceName)
			{
				
				m_minimumValue = float.MaxValue;
				m_maximumValue = float.MinValue;
				m_averagingWindow = DefaultAveragingWindow;
				m_counter = new System.Diagnostics.PerformanceCounter(categoryName, counterName, instanceName);
				m_counterValues = new List<double>();
				
			}
			
			public int AveragingWindow
			{
				get
				{
					return m_averagingWindow;
				}
				set
				{
					if (value > 0)
					{
						Interlocked.Exchange(ref m_averagingWindow, value);
					}
					else
					{
						throw (new ArgumentOutOfRangeException("AveragingWindow", "Value must be greater than 0."));
					}
				}
			}
			
			public float LastValue
			{
				get
				{
					return m_lastValue;
				}
			}
			
			public float MinimumValue
			{
				get
				{
					return m_minimumValue;
				}
			}
			
			public float MaximumValue
			{
				get
				{
					return m_maximumValue;
				}
			}
			
			public float AverageValue
			{
				get
				{
					lock(m_counterValues)
					{
						return Convert.ToSingle(Math.Common.Average(m_counterValues));
					}
				}
			}
			
			public System.Diagnostics.PerformanceCounter BaseCounter
			{
				get
				{
					return m_counter;
				}
			}
			
			public void Sample()
			{
				
				try
				{
					lock(m_counter)
					{
						Interlocked.Exchange(ref m_lastValue, m_counter.NextValue());
					}
					
					if (m_lastValue < m_minimumValue)
					{
						Interlocked.Exchange(ref m_minimumValue, m_lastValue);
					}
					if (m_lastValue > m_maximumValue)
					{
						Interlocked.Exchange(ref m_maximumValue, m_lastValue);
					}
					
					lock(m_counterValues)
					{
						m_counterValues.Add(m_lastValue);
						while (m_counterValues.Count > m_averagingWindow)
						{
							m_counterValues.RemoveAt(0);
						}
					}
				}
				catch (InvalidOperationException)
				{
					// If we're monitoring performance of an application that's not running (it was not running to begin with,
					// or it was running but it no longer running), we'll encounter an InvalidOperationException exception.
					// In this case we'll reset the values and absorb the exception.
					Reset();
				}
				
			}
			
			public void Reset()
			{
				
				Interlocked.Exchange(ref m_lastValue, 0);
				Interlocked.Exchange(ref m_minimumValue, 0);
				Interlocked.Exchange(ref m_maximumValue, 0);
				
				lock(m_counterValues)
				{
					m_counterValues.Clear();
				}
				
			}
			
			#endregion
			
			private bool disposedValue = false; // To detect redundant calls
			
			protected virtual void Dispose(bool disposing)
			{
				if (! this.disposedValue)
				{
					if (disposing)
					{
						Reset();
						lock(m_counter)
						{
							m_counter.Dispose();
						}
					}
				}
				this.disposedValue = true;
			}
			
			#region " IDisposable Support "
			// This code added by Visual Basic to correctly implement the disposable pattern.
			public void Dispose()
			{
				// Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
				Dispose(true);
				GC.SuppressFinalize(this);
			}
			#endregion
			
		}
		
	}
}

File: Source/OldCode/Archive/2005 TVA Code Library/Version/3.0/Local/Source/TVA.Core/Diagnostics/PerformanceCounter.vb
' 06/04/2007

Imports System.Threading

Namespace Diagnostics

    Public Class PerformanceCounter
        Implements IDisposable

#Region " Member Declaration "

        Private m_lastValue As Single
        Private m_minimumValue As Single
        Private m_maximumValue As Single
        Private m_averagingWindow As Integer

        Private m_counter As System.Diagnostics.PerformanceCounter
        Private m_counterValues As List(Of Double)

#End Region

#Region " Code Scope: Public "

        Public Const DefaultAveragingWindow As Integer = 120

        Public Sub New(ByVal categoryName As String, ByVal counterName As String, ByVal instanceName As String)

            MyBase.New()
            m_minimumValue = Single.MaxValue
            m_maximumValue = Single.MinValue
            m_averagingWindow = DefaultAveragingWindow
            m_counter = New System.Diagnostics.PerformanceCounter(categoryName, counterName, instanceName)
            m_counterValues = New List(Of Double)()

        End Sub

        Public Property AveragingWindow() As Integer
            Get
                Return m_averagingWindow
            End Get
            Set(ByVal value As Integer)
                If value > 0 Then
                    Interlocked.Exchange(m_averagingWindow, value)
                Else
                    Throw New ArgumentOutOfRangeException("AveragingWindow", "Value must be greater than 0.")
                End If
            End Set
        End Property

        Public ReadOnly Property LastValue() As Single
            Get
                Return m_lastValue
            End Get
        End Property

        Public ReadOnly Property MinimumValue() As Single
            Get
                Return m_minimumValue
            End Get
        End Property

        Public ReadOnly Property MaximumValue() As Single
            Get
                Return m_maximumValue
            End Get
        End Property

        Public ReadOnly Property AverageValue() As Single
            Get
                SyncLock m_counterValues
                    Return Convert.ToSingle(Math.Common.Average(m_counterValues))
                End SyncLock
            End Get
        End Property

        Public ReadOnly Property BaseCounter() As System.Diagnostics.PerformanceCounter
            Get
                Return m_counter
            End Get
        End Property

        Public Sub Sample()

            Try
                SyncLock m_counter
                    Interlocked.Exchange(m_lastValue, m_counter.NextValue())
                End SyncLock

                If m_lastValue < m_minimumValue Then Interlocked.Exchange(m_minimumValue, m_lastValue)
                If m_lastValue > m_maximumValue Then Interlocked.Exchange(m_maximumValue, m_lastValue)

                SyncLock m_counterValues
                    m_counterValues.Add(m_lastValue)
                    Do While m_counterValues.Count > m_averagingWindow
                        m_counterValues.RemoveAt(0)
                    Loop
                End SyncLock
            Catch ex As InvalidOperationException
                ' If we're monitoring performance of an application that's not running (it was not running to begin with,
                ' or it was running but it no longer running), we'll encounter an InvalidOperationException exception.
                ' In this case we'll reset the values and absorb the exception.
                Reset()
            End Try

        End Sub

        Public Sub Reset()

            Interlocked.Exchange(m_lastValue, 0)
            Interlocked.Exchange(m_minimumValue, 0)
            Interlocked.Exchange(m_maximumValue, 0)

            SyncLock m_counterValues
                m_counterValues.Clear()
            End SyncLock

        End Sub

#End Region

        Private disposedValue As Boolean = False        ' To detect redundant calls

        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposedValue Then
                If disposing Then
                    Reset()
                    SyncLock m_counter
                        m_counter.Dispose()
                    End SyncLock
                End If
            End If
            Me.disposedValue = True
        End Sub

#Region " IDisposable Support "
        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
#End Region

    End Class

End Namespace